File: src/main/Board/ChessBoard/ChessBoard.cpp
//
// Created by andre on 23/01/2022.
//


#include <iostream>
#include "ChessBoard.h"
#include "../../Utilities/random.cpp"

ChessBoard::ChessBoard() : Board(8,8){
    this-> isUserWhite = random(50);
};

void ChessBoard::createSet(bool owner){

    // Placing the pawns
    int row = owner ? 7 : 2;
    for(int i = 65; i < 72; i++){
        Position p = Position((char)i, row);
        b_pieces.insert(b_pieces.begin(), Piece(owner, "Pawn", p));
    }

    // Placing Knights, Bishops, Rooks, Queens and Kings
    if (owner){
        b_pieces.insert(b_pieces.begin(), Piece(owner, "Rook", Position('A', 8)));
        b_pieces.insert(b_pieces.begin(), Piece(owner, "Rook", Position('H', 8)));
        b_pieces.insert(b_pieces.begin(), Piece(owner, "Knight", Position('B', 8)));
        b_pieces.insert(b_pieces.begin(), Piece(owner, "Knight", Position('G', 8)));
        b_pieces.insert(b_pieces.begin(), Piece(owner, "Bishop", Position('C', 8)));
        b_pieces.insert(b_pieces.begin(), Piece(owner, "Bishop", Position('F', 8)));
        b_pieces.insert(b_pieces.begin(), Piece(owner, "Queen", Position('D', 8)));
        b_pieces.insert(b_pieces.begin(), Piece(owner, "King", Position('E', 8)));
    }
    else {
        b_pieces.insert(b_pieces.begin(), Piece(owner, "Rook", Position('A', 1)));
        b_pieces.insert(b_pieces.begin(), Piece(owner, "Rook", Position('H', 1)));
        b_pieces.insert(b_pieces.begin(), Piece(owner, "Knight", Position('B', 1)));
        b_pieces.insert(b_pieces.begin(), Piece(owner, "Knight", Position('G', 1)));
        b_pieces.insert(b_pieces.begin(), Piece(owner, "Bishop", Position('C', 1)));
        b_pieces.insert(b_pieces.begin(), Piece(owner, "Bishop", Position('F', 1)));
        b_pieces.insert(b_pieces.begin(), Piece(owner, "Queen", Position('D', 1)));
        b_pieces.insert(b_pieces.begin(), Piece(owner, "King", Position('E', 1)));
    }
}




void ChessBoard::initBoard() {;

    this-> player_turn = "white";

    /* Creation of the user set of pieces */
    createSet(true);
    /* Creation of the AI set of pieces */
    createSet(false);

}

int ChessBoard::check() {
    return 0;
}

int min(int a, int b){
    return a > b ? b : a;
}
int max(int a, int b){
    return a < b ? b : a;
}


bool ChessBoard::allowedMove(vector<Position> trace) {
    //TODO

    return true;
}


void ChessBoard::makeMove() {
    char startChar, endChar;
    int startInt, endInt;
    cout << "Insert coordinates [letter][number] x [letter][number]" <<endl;

    cin>> startChar;
    cin>> startInt;
    cin>> endChar;
    cin>> endInt;

    // TODO add integrity check for the coordinates
    Position start = Position(startChar, startInt);
    Position end = Position(endChar, endInt);

    Piece* p = this->getPiece(start);
    vector<Position>* validity = p->validMove(start, end);
    //TODO wait until the different validMove are implemented -> implement a method that check if the path of the piece is occupied
    // Note, the trace of a knight is empty because it's the only piece that can jump over other pieces



}

void printDiv(){
    cout << "   ";
    for(int i = 0; i < 8; i++){
        cout <<"-- ";
    }
    cout<<endl;
}

void ChessBoard::printBoard() {

    cout << "   ";
    for(int i = 65; i < 73; i++){
        cout <<" " <<(char)i << " ";
    }
    cout <<endl;

    for(int i = 8; i > 0; i--){
        printDiv();
        cout << i << " ";
        for(int j = 0; j < 8; j++){
            cout << "|  ";
        }
        cout << "|" << endl;

    }
    printDiv();

}


void Board::run() {




}

File: ChessApp/Models/Boards/ChessBoard.cs
using ChessApp.Models.Pieces;

namespace ChessApp.Models.Boards
{
    public class ChessBoard : Board
    {

        public int board_dimension = 8;



        public ChessBoard() : base()
        {
            initBoard();
        }


        private void initBoard()
        {

            /* Creation of the user set of pieces */
            string userPrefix;
            string pcPrefix;
            if (isUserWhite)
            {
                userPrefix = "w";
                pcPrefix = "b";
            }
            else
            {
                userPrefix = "b";
                pcPrefix = "w";
            }


            // Placing pawns
            for (int i = 0; i < 8; i++)
            {
                this.board[6,i].piece = new PawnTwoMoves(players.USER, userPrefix + "Pawn");

                this.board[1,i].piece = new PawnTwoMoves(players.AI, pcPrefix + "Pawn");

            }

            // Placing Knights, Bishops, Rooks, Queens and Kings


            this.board[0,0].piece = new Rook(players.AI, pcPrefix + "Rook");
            this.board[0,7].piece = new Rook(players.AI, pcPrefix + "Rook");
            this.board[0,1].piece = new Knight(players.AI, pcPrefix + "Knight");
            this.board[0,6].piece = new Knight(players.AI, pcPrefix + "Knight");
            this.board[0,2].piece = new Bishop(players.AI, pcPrefix + "Bishop");
            this.board[0,5].piece = new Bishop(players.AI, pcPrefix + "Bishop");
            this.board[0,3].piece = new Queen(players.AI, pcPrefix + "Queen");
            this.board[0,4].piece = new King(players.AI, pcPrefix + "King");

            this.board[7,0].piece = new Rook(players.USER, userPrefix + "Rook");
            this.board[7,7].piece = new Rook(players.USER, userPrefix + "Rook");
            this.board[7,1].piece = new Knight(players.USER, userPrefix + "Knight");
            this.board[7,6].piece = new Knight(players.USER, userPrefix + "Knight");
            this.board[7,2].piece = new Bishop(players.USER, userPrefix + "Bishop");
            this.board[7,5].piece = new Bishop(players.USER, userPrefix + "Bishop");
            this.board[7,3].piece = new Queen(players.USER, userPrefix + "Queen");
            this.board[7,4].piece = new King(players.USER, userPrefix + "King");

            for (int i = 0; i < 8; i++)
            {
                for (int j = 0; j < 8; j++)
                {
                    //this.updateCellsProtection(board[i, j]);
                }
            }


        }

        

        

        /** Return 1 if a stalemate is reached, 0 if not. */
        public int Stalemate()
        {
            return 1;
        }

        /** Return 1 if a checkmate is reached, 0 if not. */
        public bool checkMate(int status)
        {
            return true;
        }



        public new ChessBoard clone()
        {
            ChessBoard newBoard = new ChessBoard();
            for (int i = 0; i < ROW_SIZE; i++)
            {

                for (int j = 0; j < COLUMN_SIZE; j++)
                {
                    newBoard.board[i, j] = board[i, j].clone();
                }
            }
            return newBoard;
        }




    }
    
}
